internal/validators/oci/oci_validator.go

```diff
@@ -0,0 +1,142 @@
+	return &OciRuleService{
+		log: log,
+	}
+}
+
+// ReconcileOciRegistryRule reconciles an OCI registry rule from the OCIValidator config
+func (s *OciRuleService) ReconcileOciRegistryRule(rule v1alpha1.OciRegistryRule) (*vapitypes.ValidationResult, error) {
+
+	// Build the default ValidationResult for this rule
+	vr := buildValidationResult(rule, constants.OciRegistry)
+
+	// Create a new registry instance
+	reg, err := remote.NewRegistry(rule.Host)
+	if err != nil {
+		s.log.V(0).Error(err, "failed to create registry client")
+		return vr, err
+	}
+
+	httpClient := retry.DefaultClient
+	// Add cert as trust
+	if rule.Cert != "" {
+		myCert, err := base64.StdEncoding.DecodeString(rule.Cert)
+		if err != nil {
+			s.log.V(0).Error(err, "failed to decode certificate", "cert", rule.Cert)
+			return vr, err
+		}
+		caCertPool, err := x509.SystemCertPool()
+		if err != nil {
+			s.log.V(0).Error(err, "failed to get system cert pool")
+			return vr, err
+		}
+		caCertPool.AppendCertsFromPEM(myCert)
+
+		// Create TLS configuration with the certificate of the server
+		tlsConfig := &tls.Config{
+			RootCAs: caCertPool,
+		}
+		httpClient.Transport = &http.Transport{TLSClientConfig: tlsConfig}
+	}
+
+	// Setup credentials if username and password are provided
+	if rule.Auth.Username != "" && rule.Auth.Password != "" {
+		reg.Client = &auth.Client{
+			Client: httpClient,
+			Cache:  auth.DefaultCache,
+			Credential: auth.StaticCredential(rule.Host, auth.Credential{
+				Username: rule.Auth.Username,
+				Password: rule.Auth.Password,
+			}),
+		}
+	}
+
+	repoPath := rule.RepositoryPath
+
+	// Get a repository path if not present
+	ctx := context.Background()
+	if repoPath == "" {
+		var repoList []string
+		err = reg.Repositories(ctx, "", func(repos []string) error {
+			for _, repo := range repos {
+				repoList = append(repoList, repo)
+			}
+			return nil
+		})
+		if err != nil {
+			s.log.V(0).Error(err, "failed to list repositories in registry", "host", rule.Host)
+			return vr, err
+		}
+
+		if len(repoList) == 0 {
+			// no repositories in registry, not possible to run any further validations
+			return vr, nil
+		}
+
+		repoPath = repoList[0]
+	}
+
+	// Get repo from regisry
+	repo, err := reg.Repository(ctx, repoPath)
+	if err != nil {
+		s.log.V(0).Error(err, "unable to authenticate to repository", "host", rule.Host, "path", repoPath)
+		return vr, err
+	}
+
+	// Get tags on artifacts in repository
+	var tagList []string
+	err = repo.Tags(ctx, "", func(tags []string) error {
+		for _, tag := range tags {
+			tagList = append(tagList, tag)
+		}
+		return nil
+	})
+	if err != nil {
+		s.log.V(0).Error(err, "failed to pull tags from repository", "host", rule.Host, "path", repoPath)
+		return vr, err
+	}
+
+	return vr, nil
+}
+
+// buildValidationResult builds a default ValidationResult for a given validation type
+func buildValidationResult(rule ociRule, validationType string) *types.ValidationResult {
```

Comment:
The `validationType` param is a bit redundant if this method pertains to OCI rules exclusively.